File: usvm-ts/src/test/kotlin/org/usvm/samples/operators/Less.kt
package org.usvm.samples.operators

import org.jacodb.ets.model.EtsScene
import org.usvm.api.TsTestValue
import org.usvm.util.TsMethodTestRunner
import org.usvm.util.eq
import org.usvm.util.toDouble
import kotlin.test.Test

class Less : TsMethodTestRunner() {
    private val tsPath = "/samples/operators/Less.ts"

    override val scene: EtsScene = loadScene(tsPath)

    @Test
    fun testLessNumbers() {
        val method = getMethod("lessNumbers")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method,
            { a, b, r -> a.number < b.number && (r eq a) },
            { a, b, r -> b.number < a.number && (r eq b) },
            { a, b, r -> a.number == b.number && (r eq 0) },
        )
    }

    @Test
    fun testLessBooleans() {
        val method = getMethod("lessBooleans")
        discoverProperties<TsTestValue.TsBoolean, TsTestValue.TsBoolean, TsTestValue.TsBoolean>(
            method,
            { a, b, r -> !a.value && b.value && !r.value },
            { a, b, r -> a.value && !b.value && !r.value },
            { a, b, r -> a.value == b.value && !r.value },
        )
    }

    @Test
    fun testLessMixed() {
        val method = getMethod("lessMixed")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsBoolean, TsTestValue.TsNumber>(
            method,
            { a, b, r -> a.number < b.value.toDouble() && (r eq a) },
            { a, b, r -> b.value.toDouble() < a.number && (r.number == b.value.toDouble()) },
            { a, b, r -> a.number == b.value.toDouble() && (r eq 0) },
        )
    }

    @Test
    fun testLessRefs() {
        val method = getMethod("lessRefs")
        discoverProperties<TsTestValue.TsClass, TsTestValue.TsClass, TsTestValue.TsBoolean>(
            method,
        )
    }

    @Test
    fun testLessUnknown() {
        val method = getMethod("lessUnknown")
        discoverProperties<TsTestValue.TsUnknown, TsTestValue.TsUnknown, TsTestValue.TsBoolean>(
            method,
        )
    }
}

File: usvm-ts/src/test/resources/samples/operators/Less.ts
class Less {
    lessNumbers(a: number, b: number): number {
        if (a < b) {
            return a;
        }

        if (b < a) {
            return b;
        }

        return 0; // Return 0 if they are equal
    }

    lessBooleans(a: boolean, b: boolean): boolean {
        if (a < b) {
            return a;
        }

        if (b < a) {
            return b;
        }

        return false; // Return false if they are equal
    }

    lessMixed(a: number, b: boolean): number {
        if (a < b) {
            return a;
        }

        if (b < a) {
            return b ? 1 : 0; // Convert boolean to number
        }

        return 0; // Return 0 if they are equal
    }

    lessRefs(a: object, b: object): boolean {
        if (a < b) {
            return true; // Reference comparison, not value
        }

        if (b < a) {
            return false; // Reference comparison, not value
        }

        return false; // Return false if they are equal
    }

    lessUnknown(a, b): boolean {
        if (a < b) {
            return true; // Reference comparison, not value
        }

        if (b < a) {
            return false; // Reference comparison, not value
        }

        return false; // Return false if they are equal
    }
}


